### DISCUSSION TITLE
Environment var in redirect URL?

### DISCUSSION THREAD

**Discussion by guifry** [vercel/vercel#4351] [General] [answered]
Say environment variable `BACKEND_URL=back.com`

I want to use it in my redirects as follow, in my now.json:
```json
{
    "redirects": [
        { "source": "/api/(.*)", "destination": "BACKEND_URL/$1", "statusCode": 200 }
    ]
}
```

Then I try to replace it on the fly in the build command:
`s|BACKEND_URL|${BACKEND_URL}|g" now.json && yarn build`

But unfortunately it can't find `now.json`, although I can see it as part of the source files once the deploy has failed.

How to achieve this?

**Comment by styfle** [comment, marked as answer]
Hi @JackPoteYes 

Environment Variables can't be used in config, only in code.

Also, a 200 status code is not a valid [redirect](https://vercel.com/docs/configuration#project/redirects) so it's unlikely that a user agent would follow the redirect without something in the 300 range.

Maybe you want to use a [rewrite](https://vercel.com/docs/configuration#project/rewrites) instead?

**Comment by guifry** [comment]
Hi @styfle 

Thanks for your input!

Using rewrite won't fix the original issue I'm affraid. The `BACKEND_URL` depends on the environment the app is deployed in (i.e staging backend url for staging frontend etc.) and therefore should be dynamic. 

I guess I'll have to look at alternative solutions.

**Comment by styfle** [comment]
You could use [Next.js](https://nextjs.org) which provides a `next.config.js` that can read environment variables as defined in [RFC 9081](https://github.com/zeit/next.js/issues/9081).

```
npm install next@canary react react-dom
```

```js
// next.config.js
module.exports = {
  experimental: {
    async redirects() {
      return [{
        source: "/api/:path*",
        destination: `${process.env.BACKEND_URL}/:path*`,
        permanent: false,
      }]
    }
  }
};
```

**Comment by guifry** [comment]
Hi @styfle that is an interesting feature.
However our project is too advanced in production to consider a switch of technology.
We found another solution that we perform on our backend server and had nothing to do with Vercel.

Thank you anyway!

**Comment by guifry** [comment]
TDLR: Dynamic redirect and rewrites that depend on the environment are not possible.
A solution might be to use Next.JS with on the fly declaration of the redirects.

**Comment by peaceful-james** [comment]
I think this actually is possible.
If you are using create-react-app then you just need to preface your env var name with `REACT_APP_`. See here: https://create-react-app.dev/docs/adding-custom-environment-variables/

I am currently doing this with a websocket URL env var. I have named it `REACT_APP_WS_URL` and here is how I use it.

I have 2 different `vercel` files in the project:

1. `vercel.staging.json` which has this section:
```
  "build": {
    "env": {
      "REACT_APP_WS_URL": "wss:staging.my-backend.com/socket"
    }
  }
```

2. `vercel.live.json` which has this section:
```
  "build": {
    "env": {
      "REACT_APP_WS_URL": "wss:my-backend.com/socket"
    }
  }
```

I deploy to them with either of these commands:

```
vercel deploy --prod -A vercel.staging.json
```

```
vercel deploy --prod -A vercel.live.json
```

and in my code I can access `process.env.REACT_APP_WS_URL` anywhere.

I have not tried doing this with the Vercel dashboard env vars but it might be worth trying your original approach except rename your env var to `REACT_APP_BACKEND_URL`.

Note: my deployment commands only work when I don't assign domains to the project. If I assign domains to a project, they are automatically used for ALL `--prod` deploys, no matter what is in my `alias` field in the json config file.

I hope this helps somebody.
